### PR TITLE
Book Progress Tracking with Total Chapters and Usernames

### PR DESCRIPTION
**Pull request by Shromm-Gaind**
### Pull Request Description

- **Added `totalPages` to `Book` model:**
  - Enhanced the `Book` class to include a `totalPages` attribute for tracking the total number of pages.
  
- **Integrated `totalPages` in Book Progress UI:**
  - Updated the UI to display both current page and total pages for each book.
  
- **Added user-specific progress tracking:**
  - Implemented logic to display each user's progress as "Username is at Page X/Y" in the progress list.
  
- **Updated `BookProgressService` and DAOs:**
  - Modified `BookProgressService` to retrieve and display usernames using `UserDAO` for better progress tracking.
  
- **Refactored `initialize` method:**
  - Enhanced the `BookProgressService` initialization to accept both `BookProgressDAO` and `UserDAO` for managing book progress and user information.

### CODE REVIEW

**Comment by Shromm-Gaind** [comment]
I also made quite a few changes to my section. My logic and changes is as follows:

### Changes to `BookDAO`:

- **Generalized Query Method (`getBooksFromQuery`)**:
  - **What changed**: Combined the logic for fetching books based on various criteria into a single method (`getBooksFromQuery`).
  - **Justification**: This eliminates repetitive code.

- **Used Try-with-Resources**:
  - **What changed**: Refactored database connection handling to use try-with-resources.
  - **Justification**: This ensures that resources like `ResultSet`, `PreparedStatement`, and `Connection` are properly closed.

- **Logging Instead of `printStackTrace()`**:
  - **What changed**: Replaced `printStackTrace()` calls with proper logging using `Logger`.
  - **Justification**: This provides better error handling. It was also just cleaner

---

### Changes to `BookProgressService`:

- **Cache User Data**:
  - **What changed**: Cached user data in `getFormattedProgressForBook()` to avoid repeated calls to `userAO.findUserById()`.
  - **Justification**: This improves performance by reducing repeated database lookups.

- **Remove Redundant Logic in `saveBookProgress()`**:
  - **What changed**: Unified the logic for adding and updating book progress into a single method.
  - **Justification**: This reduces redundant code.

- **Use `Optional` to Avoid Null Checks**:
  - **What changed**: Leveraged `Optional` to handle potential null values in `getBookProgress()`.
  - **Justification**: This improves code readability and reduces the need for explicit null checks. I also just don't like null checks

---

### Changes to `BookProgressController`:

- **Minimize Calls to Services**:
  - **What changed**: Reduced repeated calls to `BookService.getInstance().getSelectedBook()` and `LoginService.getCurrentUser()` by storing the results in local variables.
  - **Justification**: This reduces redundant operations and improves performance by minimizing service calls.

- **Update UI More Efficiently**:
  - **What changed**: Combined updates for labels and the progress bar to refresh the UI more efficiently.
  - **Justification**: This minimizes unnecessary UI refreshes.

---

File: src/main/java/com/bookclub/dao/BookDAO.java [hunk 1]
```diff
@@ -6,189 +6,160 @@
...
+        return books;
     }
 
-    private List<Book> getListFromStatement(PreparedStatement statement) {
+    // Fetch the list of books based on a prepared statement
+    private List<Book> getListFromStatement(PreparedStatement statement) throws SQLException {
         List<Book> books = new ArrayList<>();
-        try {
-            ResultSet resultSet = statement.executeQuery();
+        try (ResultSet resultSet = statement.executeQuery()) {
```
Comment (jmccook96) [inline src/main/java/com/bookclub/dao/BookDAO.java:53]: What's up with everyone changing the try foramt we're using...

File: src/main/java/com/bookclub/dao/BookDAO.java [hunk 2]
```diff
@@ -6,189 +6,160 @@
...
+        } catch (SQLException e) {
+            LOGGER.log(Level.SEVERE, "Error fetching books", e);
         }
-            return null;
+        return books;
     }
 
-    private List<Book> getListFromStatement(PreparedStatement statement) {
+    // Fetch the list of books based on a prepared statement
+    private List<Book> getListFromStatement(PreparedStatement statement) throws SQLException {
```
Comment (jmccook96) [inline src/main/java/com/bookclub/dao/BookDAO.java:51]: Why are we publicly throwing this exception now? Doesn't this mean the IAO & MAO has to be updated to throw an exception that they don't even know exists?

File: src/main/java/com/bookclub/dao/BookDAO.java [hunk 3]
```diff
@@ -6,189 +6,160 @@
...
+    // Generalized method to fetch a list of books with different queries
+    private List<Book> getBooksFromQuery(String query, String param) {
+        List<Book> books = new ArrayList<>();
+        try (PreparedStatement statement = dbManager.getConnection().prepareStatement(query)) {
+            if (param != null) {
+                statement.setString(1, param);
+            }
+            books = getListFromStatement(statement);
+        } catch (SQLException e) {
+            LOGGER.log(Level.SEVERE, "Error fetching books", e);
```
Comment (jmccook96) [inline src/main/java/com/bookclub/dao/BookDAO.java:45]: Why have we changed logging format too? While better, strange change to spontaneously decide to implement...

File: src/main/java/com/bookclub/dao/BookDAO.java [hunk 4]
```diff
@@ -6,189 +6,160 @@
...
     public boolean updateBook(Book book) {
-        try {
-            PreparedStatement statement = dbManager.getConnection().prepareStatement("UPDATE Books SET bookTitle = ?, bookAuthor = ? WHERE bookID = ?");
+        String query = "UPDATE Books SET bookTitle = ?, bookAuthor = ?, bookGenre = ?, totalPages = ? WHERE bookID = ?";
+        try (PreparedStatement statement = dbManager.getConnection().prepareStatement(query)) {
             statement.setString(1, book.getTitle());
             statement.setString(2, book.getAuthor());
-            statement.setInt(3, book.getId());
+            statement.setString(3, book.getGenre());
+            statement.setInt(4, book.getTotalPages());
```
Comment (jmccook96) [inline src/main/java/com/bookclub/dao/BookDAO.java:143]: Total pages being the measure feels dangerous. I have 3 different printings of Dantes Inferno, each with wildly different page counts. Chapter count might have been a better idea, and also an easier meta metric to measure against in general, but maybe too late now.

File: src/main/java/com/bookclub/dao/BookDAO.java [hunk 5]
```diff
@@ -6,189 +6,160 @@
...
-
+        } catch (SQLException e) {
+            LOGGER.log(Level.SEVERE, "Error deleting book", e);
+            return false;
         }
         return true;
     }
-
-
-    private void createTable() {
```
Comment (jmccook96) [inline src/main/java/com/bookclub/dao/BookDAO.java:178]: It feels weird to have moved all the meta populating functions (that is all the ones not publicly listed) are the ones first in this file. Very backwards to standard practice.

File: src/main/java/com/bookclub/mao/BookProgressMAO.java [hunk 6]
```diff
@@ -53,9 +53,18 @@ public boolean addBookProgress(BookProgress bookProgress) {
 
     @Override
     public boolean updateBookProgress(BookProgress bookProgress) {
-        return bookProgressList.set(bookProgressList.indexOf(findBookProgressByBookAndUser(bookProgress.getBookId(), bookProgress.getUserId())), bookProgress) != null;
+        int index = bookProgressList.indexOf(findBookProgressByBookAndUser(bookProgress.getBookId(), bookProgress.getUserId()));
```
Comment (jmccook96) [inline src/main/java/com/bookclub/mao/BookProgressMAO.java:56]: Is this a bug? I don't understand why this change is here, thought `set` was a safe function present or absent data

File: src/main/java/com/bookclub/model/Book.java [hunk 7]
```diff
@@ -148,6 +153,14 @@ public void setAuthor(String author) {
         this.author = author;
     }
 
+    public int getTotalPages() {
+        return totalPages;
+    }
+
+    public void setTotalPages(int totalPages) {
+        this.totalPages = totalPages;
```
Comment (jmccook96) [inline src/main/java/com/bookclub/model/Book.java:161]: -1 shouldn't be considered a valid number of pages

File: src/main/java/com/bookclub/service/BookProgressService.java [hunk 8]
```diff
@@ -2,58 +2,89 @@
...
             instance = new BookProgressService();
         }
         instance.bookProgressAO = bookProgressAO;
+        instance.userAO = userAO;
     }
 
-    public BookProgress getBookProgress(Book book, User user) {
-        return bookProgressAO.findBookProgressByBookAndUser(book.getId(), user.getId());
+    public Optional<BookProgress> getBookProgress(Book book, User user) {
+        return Optional.ofNullable(bookProgressAO.findBookProgressByBookAndUser(book.getId(), user.getId()));
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookProgressService.java:34]: What is the point of declaring as optional like this?

File: src/main/java/com/bookclub/service/BookProgressService.java [hunk 9]
```diff
@@ -2,58 +2,89 @@
...
+                .map(bookProgress -> {
+                    // Update existing progress
+                    bookProgress.setPageNumber(pageNumber);
+                    return bookProgressAO.updateBookProgress(bookProgress);
+                })
+                .orElseGet(() -> {
+                    // Create new progress
+                    BookProgress newProgress = new BookProgress(book.getId(), user.getId(), pageNumber);
+                    return bookProgressAO.addBookProgress(newProgress);
+                });
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookProgressService.java:69]: God I hate this from a formatting perspective

File: src/main/java/com/bookclub/service/BookService.java [hunk 10]
```diff
@@ -48,35 +48,22 @@ public void setSelectedBook(Book book) {
         selectedBook.set(book);
     }
 
+    // Optimized search logic: combined filters using a single stream
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookService.java:51]: Your AI is showing in your comments

File: src/main/java/com/bookclub/service/BookService.java [hunk 11]
```diff
@@ -48,35 +48,22 @@ public void setSelectedBook(Book book) {
...
-            filteredBooks = filteredBooks.stream()
-                    .filter(book -> book.getGenre().equalsIgnoreCase(selectedGenre))
-                    .collect(Collectors.toList());
-        }
-        return filteredBooks;
+        return filteredBooks.stream()
+                .filter(book -> (searchQuery.isEmpty() || book.getTitle().toLowerCase().contains(searchQuery.toLowerCase())
+                        || book.getAuthor().toLowerCase().contains(searchQuery.toLowerCase())))
+                .filter(book -> selectedGenre.equalsIgnoreCase("All") || book.getGenre().equalsIgnoreCase(selectedGenre))
+                .collect(Collectors.toList());
```
Comment (jmccook96) [inline src/main/java/com/bookclub/service/BookService.java:59]: This is such an unnatural change compared to everything else that has happened in this repository, and is out of scope of what you were working on.

File: src/test/java/com/example/BookProgressServiceTest.java [hunk 12]
```diff
@@ -23,23 +24,23 @@ public class BookProgressServiceTest {
     @BeforeEach
     public void setUp() {
         bookProgressMAO = new BookProgressMAO();
-        book = new Book(1, "TestTitle", "TestAuthor", "TestGenre");
+        book = new Book(1, "TestTitle", "TestAuthor", "TestGenre", 100);
         user = new User(1, "testUser", "testPassword", "Test Name", "test@example.com");
-        BookProgressService.initialize(bookProgressMAO);
+        BookProgressService.initialize(bookProgressMAO, new UserDAO());  // Initialize with both MAO and UserDAO
```
Comment (jmccook96) [inline src/test/java/com/example/BookProgressServiceTest.java:29]: A DAO? In a test suite?

An object that we actively don't test because it's dangerous and unreliable in automated testing- used for testing purposes of a different service?

Comment (jmccook96) [reply src/test/java/com/example/BookProgressServiceTest.java:29]: I am shocked and scared that the maven automated testing didn't fail on this

File: src/test/java/com/example/BookProgressServiceTest.java [hunk 13]
```diff
@@ -23,23 +24,23 @@ public class BookProgressServiceTest {
...
         bookProgressService = BookProgressService.getInstance();
     }
 
     @Test
     void testGetBookProgress_NoProgressExists() {
-        BookProgress actual = bookProgressService.getBookProgress(book, user);
+        BookProgress actual = bookProgressService.getBookProgress(book, user).orElse(null);
```
Comment (jmccook96) [inline src/test/java/com/example/BookProgressServiceTest.java:35]: What is the point of this orElse(null)? What's changed that has made this necessary? Isn't this just returning null | null ?

File: src/main/java/com/bookclub/mao/BookProgressMAO.java (hunk 6 above)
Comment (Shromm-Gaind) [reply src/main/java/com/bookclub/mao/BookProgressMAO.java:56]: in the original code, 

the method relies on set to update the element at a given index, however if the element is not found, index of,m would return -1 and calling set with an invalid index would result in an index out of bounds exception 

What I ended up doing was, 

By checking index >= 0, I ensure that set is only called when the item exists. 

So yes it was a bug fix.

File: src/main/java/com/bookclub/service/BookProgressService.java (hunk 8 above)
Comment (Shromm-Gaind) [reply src/main/java/com/bookclub/service/BookProgressService.java:34]: This is good practice no... null seems idk. I can change it back if you want.

I believe using Optional here is a valid practice, as it explicitly handles cases where findbookprogressbybookanduser may return null. For example, if the bookprogress for the given book and user is not found, Optional.ofNullable allows us to avoid potential NullPointerExceptions and ensures that the absence of a result is handled "nicely" (I can't think of the word).

File: src/test/java/com/example/BookProgressServiceTest.java (hunk 13 above)
Comment (Shromm-Gaind) [reply src/test/java/com/example/BookProgressServiceTest.java:35]: I added the orElse(null) because it is used to handle the optional returned by getbookprogress. 

Previously when the getbookprogress method didn't find any progress it would have returned null, so we would check in this case weather the result is null to confirm that no progress exists...

With the optional<bookprogress>, if there's no book progress, this would be empty and wouldn't hold any value.

The orElse(null) would therefore allow us to get back to the previous behavior of returning null when no progress is found.

File: src/main/java/com/bookclub/dao/BookDAO.java (hunk 1 above)
Comment (Shromm-Gaind) [reply src/main/java/com/bookclub/dao/BookDAO.java]: Book DAO has been changed back

File: src/test/java/com/example/BookProgressServiceTest.java (hunk 12 above)
Comment (Shromm-Gaind) [reply src/test/java/com/example/BookProgressServiceTest.java]: I changed it to initialize with UserMAO.

File: src/main/java/com/bookclub/service/BookProgressService.java (hunk 8 above)
Comment (jmccook96) [reply src/main/java/com/bookclub/service/BookProgressService.java:60]: I'm cool with optional :)

Is the right practice for a public function, you are right.

**Comment by jmccook96** [comment]
Am I going to be able to convince you in this PR to pivot the wording and names to reflect Chapter Progress instead of Book Progress?

**Review by jmccook96 — APPROVED** [review]
